### PR TITLE
Added NoUselessReturnFixer

### PR DESCRIPTION
**Pull request by SpacePossum**
replaces:
https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/1591

### CODE REVIEW

**Comment by SpacePossum** [comment]
Diff. of Fixer behavior is minimal, biggest change is the clear and merge logic added to Tokens.
This change will great an option to start finishing some other PRs.

File: Symfony/CS/Tokenizer/Tokens.php [hunk 1]
```diff
@@ -1403,4 +1403,45 @@ public function isMonolithicPhp()
...
+
+        for ($from = $index - 1; $from > 0; --$from) {
+            if (!$this[$from]->isEmpty()) {
+                break;
+            }
+        }
+
+        if ($this[$from]->isWhitespace()) {
+            $this[$from]->setContent($this[$from]->getContent().$this[$till]->getContent());
+        } elseif ($this[$from + 1]->isEmpty('')) {
```
Comment (keradus) [inline Symfony/CS/Tokenizer/Tokens.php]: I will need to take a deeper look into this method.
Here just one picky comment - `isEmpty` does not have any parameter

File: Symfony/CS/Tokenizer/Tokens.php [hunk 2]
```diff
@@ -1403,4 +1403,45 @@ public function isMonolithicPhp()
...
+            $this[$from + 1]->override(array(T_WHITESPACE, $this[$till]->getContent(), $this[$from + 1]->getLine()));
```
Comment (keradus) [inline Symfony/CS/Tokenizer/Tokens.php]: NOTE to self: need to be extra careful during merging to 2.x

File: Symfony/CS/Tokenizer/Tokens.php (hunk 1 above)
Comment (SpacePossum) [inline Symfony/CS/Tokenizer/Tokens.php]: ah yes, thanks! (leftover from `->equals('')` to `->isEmpty()` :(  )

Comment (SpacePossum) [inline Symfony/CS/Tokenizer/Tokens.php]: fixed, ping @keradus

File: Symfony/CS/Tests/Fixtures/Integration/priority/empty_return,useless_return,whitespacy_lines,extra_empty_lines.test [hunk 3]
```diff
@@ -0,0 +1,45 @@
+--TEST--
```
Comment (keradus) [inline Symfony/CS/Tests/Fixtures/Integration/priority/empty_return,useless_return,whitespacy_lines,extra_empty_lines.test]: Please move this test to misc.
Fixers in `misc` could check priority as well.
`priority` dir is one to follow relations between 2 fixers defined in `FixerTest.php`

File: Symfony/CS/Tokenizer/Tokens.php [hunk 4]
```diff
@@ -1404,4 +1404,45 @@ public function isMonolithicPhp()
...
+    public function clearTokenAndMergeSurroundingWhitespace($index)
+    {
+        $count = count($this);
+        $this[$index]->clear();
+        if ($index === $count - 1) {
+            return;
+        }
+
+        // We are looking for the following sequence:
+        // [w][c](0..N)[index][c](0..N)[w]
```
Comment (keradus) [inline Symfony/CS/Tokenizer/Tokens.php]: Sorry, I tried to understand it few times and still, I have no idea what it is...

Please, consider following change: https://github.com/SpacePossum/PHP-CS-Fixer/pull/10

**Comment by keradus** [comment]
Please, fix Travis

**Comment by keradus** [comment]
3 comments, all the rest look OK for me

File: Symfony/CS/Tokenizer/Tokens.php (hunk 4 above)
Comment (SpacePossum) [inline Symfony/CS/Tokenizer/Tokens.php]: I remember ^-^
this sequence:
[1 white space token][any number of cleared tokens][token at given index][any number of cleared tokens][1 white space token]

not sure why, I'm checking your PR now :)

**Comment by GrahamCampbell** [comment]
Needs squashing to 1 commits per author.

**Comment by keradus** [comment]
:+1: (after squashing)

File: Symfony/CS/Fixer/Contrib/UselessReturnFixer.php [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+final class UselessReturnFixer extends AbstractFixer
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function fix(\SplFileInfo $file, $content)
+    {
+        $tokens = Tokens::fromCode($content);
+
+        $functionTokens = $tokens->findGivenKind(T_FUNCTION);
```
Comment (GrahamCampbell) [inline Symfony/CS/Fixer/Contrib/UselessReturnFixer.php]: findGivenKind should be avoided

Comment (GrahamCampbell) [inline Symfony/CS/Fixer/Contrib/UselessReturnFixer.php]: @keradus Can we actually mark that method as deprecated in 1.12?

Comment (SpacePossum) [inline Symfony/CS/Fixer/Contrib/UselessReturnFixer.php]: what did I miss, there are 40+ uses on 1.12?

Comment (keradus) [inline Symfony/CS/Fixer/Contrib/UselessReturnFixer.php]: we have valid cases for that method. It's just not sth to be used everywhere

**Comment by SpacePossum** [comment]
>(after squashing)

squashed to 1 commit by author, is this ok?

File: Symfony/CS/Fixer/Contrib/UselessReturnFixer.php (hunk 5 above)
Comment (keradus) [inline Symfony/CS/Fixer/Contrib/UselessReturnFixer.php]: @SpacePossum it is not that the method should not be used at all.
But here you don't need it - check token kind inside loop and you won't need to iterate twice through every single token

**Comment by keradus** [comment]
Please fix your commit log to sth more meaningful ;)

**Comment by keradus** [comment]
ping @SpacePossum

**Comment by keradus** [comment]
```
git checkout 538dd8d
git pick --no-commit d999763
git commit --amend -m "Added UselessReturnFixer"
git pick c84039ba4d4a20389099666dcc2a0315fa8dec9e
git push origin 1_12_clear_tokens -f
```

**Comment by SpacePossum** [comment]
thanks for hints, I did a little detour but I think this should be OK-ish now

**Comment by keradus** [comment]
btw, just a thought, what about `no_useless_return` ? as in fact we are removing the statement ?

**Comment by SpacePossum** [comment]
>no_useless_return

:+1: will rename and do the squashing thing again later today or maybe tomorrow morning.
@ all let me know if there are other things to be fixed in this PR as well :)

**Comment by keradus** [comment]
?

**Comment by SpacePossum** [comment]
github auto closes the PR if I reset the branch back to 0 commits, than force push. After that I do the picking and merging and have to re-open the PR by hand. Not sure how to prevent this :/

**Comment by keradus** [comment]
don't push before picking.

**Comment by keradus** [comment]
You need to regenerate readme

**Comment by SpacePossum** [comment]
d'oh, no time for the squashing, will do later tonight

**Comment by keradus** [comment]
I am still wondering how you squash... a lot of time and PR closing/opening.
It's just few commands ([eg](https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/1705#issuecomment-181000024))

**Comment by SpacePossum** [comment]
on the branch, no changes (of interest)
- git reset HEAD~3
- rm Symfony -rf
- git reset --hard
- git cherry-pick [fist commit]
- git cherry-pick --no-commit [3rd commit]
- git pick 2nd commit (yours)
error: Your local changes would be overwritten by cherry-pick.
- git reset README..
- git co READ..
- git chery-pick again
- git push -f origin

**Comment by SpacePossum** [comment]
and readme is wrong again...

**Comment by keradus** [comment]
@SpacePossum , just copy-execute:
```bash
D:\GIT\fork\PHP-CS-Fixer (2.x-async_linting)
λ git co 9656a82af22011ada04629618c82e96d3e902f02
Note: checking out '9656a82af22011ada04629618c82e96d3e902f02'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by performing another checkout.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -b with the checkout command again. Example:

  git checkout -b new_branch_name

HEAD is now at 9656a82... Added NoUselessReturnFixer

D:\GIT\fork\PHP-CS-Fixer ((detached from 9656a82))
λ git cherry-pick --no-commit cdfe52d18e87e1466054fd154f98f546270a3c6f

D:\GIT\fork\PHP-CS-Fixer ((detached from 9656a82))
λ git commit --amend --no-edit
[detached HEAD 9e4d4e3] Added NoUselessReturnFixer
 Author: Possum <possumfromspace@gmail.com>
 19 files changed, 684 insertions(+), 20 deletions(-)
 create mode 100644 Symfony/CS/Fixer/Contrib/NoUselessReturnFixer.php
 create mode 100644 Symfony/CS/Tests/Fixer/Contrib/NoUselessReturnFixerTest.php
 create mode 100644 Symfony/CS/Tests/Fixtures/Integration/misc/empty_return,useless_return,whitespacy_lines,extra_empty_lines.test
 create mode 100644 Symfony/CS/Tests/Fixtures/Integration/priority/duplicate_semicolon,no_useless_return.test
 create mode 100644 Symfony/CS/Tests/Fixtures/Integration/priority/empty_return,no_useless_return.test
 create mode 100644 Symfony/CS/Tests/Fixtures/Integration/priority/no_useless_return,extra_empty_lines.test
 create mode 100644 Symfony/CS/Tests/Fixtures/Integration/priority/no_useless_return,return.test
 create mode 100644 Symfony/CS/Tests/Fixtures/Integration/priority/no_useless_return,whitespacy_lines.test

D:\GIT\fork\PHP-CS-Fixer ((detached from 9656a82))
λ git cherry-pick 349f754215b201ab505d20e2ed86f46dcbf7ca0d
[detached HEAD bb4bc6b] Tokens::clearTokenAndMergeSurroundingWhitespace - code grooming
 4 files changed, 36 insertions(+), 41 deletions(-)

D:\GIT\fork\PHP-CS-Fixer ((detached from 9656a82))
λ git branch -D 1_12_clear_tokens
error: branch '1_12_clear_tokens' not found.

D:\GIT\fork\PHP-CS-Fixer ((detached from 9656a82))
λ git checkout -b 1_12_clear_tokens
Switched to a new branch '1_12_clear_tokens'

D:\GIT\fork\PHP-CS-Fixer (1_12_clear_tokens)
λ git push origin 1_12_clear_tokens -f
Counting objects: 229, done.
Delta compression using up to 2 threads.
Compressing objects: 100% (49/49), done.
Writing objects: 100% (51/51), 9.62 KiB | 0 bytes/s, done.
Total 51 (delta 41), reused 0 (delta 0)
To git@github.com:keradus/PHP-CS-Fixer.git
 * [new branch]      1_12_clear_tokens -> 1_12_clear_tokens
```

**Comment by SpacePossum** [comment]
thanks so much @keradus !
all done, lets see what travis and co. think :)

**Comment by keradus** [comment]
Thank you @SpacePossum.

**Comment by SpacePossum** [comment]
Thanks all for reviewing :)
Thanks @keradus for the assists with the git stuff and for merging, this opens up a lot of new options for new and better fixers :D

**Comment by Soullivaneuh** [comment]
@SpacePossum Not sure of that, but should not be this fixer marked as risky since PHP 7.1 and the void/null return difference?

**Comment by SpacePossum** [comment]
@Soullivaneuh thanks for the heads up. 
At the time of writing the `void` type was only a thing talked about, but now we should indeed update the fixer and/or mark it risky. 

More discussion in detail about what people think would be a good way;
https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/1638#issuecomment-205916520
https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/1916

**Pull request merged**